File: Arduino/sketch/servo.cpp
#include "servo.h"

extern char jsonOut[256];
extern int written;


Servo servos[MAX_PINS];

void handleServoRequest(JsonObject& root)
{
  const char* method = root["servo"];
  int pin = root["pin"];
  StaticJsonBuffer<200> responseJsonBuffer;
  JsonObject& response = responseJsonBuffer.createObject();
  response["id"] = root["id"];

  if (strcmp(method, "attach") == 0) {
    response["deprecated"] = (bool)true;
    if (!servos[pin].attached()) {
      servos[pin].attach(pin);
    }
  } else if (strcmp(method, "detach") == 0) {
    if (servos[pin].attached()) {
      servos[pin].detach();
    }
  } else if (strcmp(method, "write") == 0) {
    int degrees = root["degrees"];

    if (!servos[pin].attached()) {
      servos[pin].attach(pin);
    }
    
    if (servos[pin].read() != degrees) {
      servos[pin].write(degrees);
    }
   
  } else {
    response["success"] = (bool)false;
  }
  
  response["success"] = (bool)true;

  written = response.printTo(jsonOut, 255);
  jsonOut[written] = '\n';
  written++;
  jsonOut[written] = '\0';
  reply(jsonOut, written);
}

File: tests/servo.py
import time
from belzetest import *

test = BelzeTest()
test.openport()


print("Servo 0 degrees");
test.write('{"id": 100, "servo":"write", "degrees":0, "pin":6}');
time.sleep(2);

print("Servo 10 degrees");
test.write('{"id": 101, "servo":"write", "degrees":10, "pin":6}');
time.sleep(2);

print("Servo 90 degrees");
test.write('{"id": 102, "servo":"write", "degrees":90, "pin":6}');
time.sleep(2);

print("Servo 180 degrees");
test.write('{"id": 103, "servo":"write", "degrees":180, "pin":6}');
time.sleep(2);

print("Servo 60 degrees");
test.write('{"id": 104, "servo":"write", "degrees":60, "pin":6}');
time.sleep(0.5);

print("Servo 60 degrees");
test.write('{"id": 104, "servo":"write", "degrees":60, "pin":6}');
time.sleep(0.5);

print("Servo 60 degrees");
test.write('{"id": 104, "servo":"write", "degrees":60, "pin":6}');


test.closeport();
